core/runtime/src/main/java/io/quarkus/runtime/logging/CategoryBuildTimeConfig.java

```diff
@@ -0,0 +1,23 @@
+package io.quarkus.runtime.logging;
+
+import io.quarkus.runtime.annotations.ConfigGroup;
+import io.quarkus.runtime.annotations.ConfigItem;
+
+@ConfigGroup
+public class CategoryBuildTimeConfig {
+    /**
+     * Specifies whether <code>isTraceEnabled</code> for a given category will return true or false for a given category.
+     * This check is computed at build time to provide better hints on code paths that won't be used at runtime.
+     *
+     * It's common to see <code>isTraceEnabled</code> calls in sections that include building complex or expensive messages.
+     * By using <code>isTraceEnabled</code> checks, users can double check that trace enabled before computing such a message.
+     * Otherwise, complex messages could be constructed but then not used by the logging library if the runtime log level is not trace.
+     * This build time option controls whether `isTraceEnabled` returns true or false at build time and makes that a constant at runtime,
+     * irrespective of the runtime log level.
+     *
+     * Calls to <code>trace()</code> not preceded by <code>isTraceEnabled</code> are still governed by the runtime log level.
```

Comment:
This is a bit odd in terms of semantics, no? Wouldn't it be far easier for users to understand this feature if the focus was about the trace cathegory as a whole not being something which can be controlled at runtime?

I love the potential of your patch, but in the current shape I'm a bit concerned about it exposing much complexity.

Reply:
Yeah, it feels a bit odd. I already hinted at this issue [here](https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/All.20log.20levels.20enabled/near/214597958), but @dmlloyd opposed the idea that build time trace configuration should delete all trace calls.

To make this a fully build time feature, aside of what I did in this PR, you'd need to remove `quarkus.log.category.<...>.level=TRACE` option, meaning the ability to set TRACE as a runtime log level.

Reply:
Note that as @dmlloyd said, an option can only either be build time or runtime.

Reply:
> To make this a fully build time feature, aside of what I did in this PR, you'd need to remove quarkus.log.category.<...>.level=TRACE option, meaning the ability to set TRACE as a runtime log level.

Right, I can see how people would find that odd as well. But personally I think that would be better: this kinda needs a re-definition of a Level to be special.

Ideally the format of configuration could be more specific; allowing the user to choose which runtime configuration elements are made available via an enumeration - I proposed this on the list in the past.

My past proposal was expressed in the area of databases; so in the build time configuration one could, for example, leave MySQL out:

supported_databases={H2,Oracle,PostgreSQL}

then at runtime one has to choose ONE *among* the enumeration.

Similarly for log levels one could have

`<    >.level={TRACE,DEBUG,INFO}`

or choose 

`<    >.level={DEBUG,INFO}`

Which would make the `TRACE` option out, exluding it from the valid runtime options.

IMO it has the following benefits:

1. it's clear what the end user is opting for
2. people who prefer having more options at runtime can have their way if they want to
3. Clean, precise separation of build-time vs runtime configuration

Reply:
We had effectively this feature in the `min-level` configuration.  Filtering out arbitrary levels doesn't seem as specifically useful as setting a single bound (and might lead to some confusion).  It certainly would be more complex to implement.

Reply:
@dmlloyd raises a valid point here. Although it sounds useful, it also opens up to some potential misuse, e.g. what if the user sets `TRACE,WARN`, do you really only want `TRACE` and `WARN` and not `DEBUG`, `INFO`? If you want to handle ranges, you could add support for `..`...

`min-level` might make more sense, from level X upward. Why was that removed? Was the name confusing? Do you need the flexibility of the enumeration?

Reply:
I agree, `min-level` looks good. I don't know why it was removed.

Reply:
Looking at `LogConfig`, `min-level` seemed to be a runtime option. It got deprecated in #10672, but at first glance I don't see specific reasons on why it got deprecated.

Reply:
It _was_ a build time option originally.  It was removed because nobody used it or showed much interest in it.

Reply:
I'm working on a update to the PR.